core/src/main/java/com/datastrato/gravitino/lock/TreeLockNode.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.lock;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * TreeLockNode is a node in the tree lock; all tree lock nodes will be assembled to a tree
+ * structure, which corresponds to the resource path like name identifier space.
+ *
+ * <p>Each node will have a read-write lock to protect the node. The node will also have a map to
+ * store the children. For more, please refer to {@link TreeLock}.
+ */
+public class TreeLockNode {
+  public static final Logger LOG = LoggerFactory.getLogger(TreeLockNode.class);
+  private final String name;
+  private final ReentrantReadWriteLock readWriteLock;
+  @VisibleForTesting final Map<String, TreeLockNode> childMap;
+  private final LockManager lockManager;
+
+  // The reference count of this node. The reference count is used to track the number of the
+  // TreeLocks that are using this node. If the reference count is 0, it means that no TreeLock is
+  // using this node, and this node can be removed from the tree.
+  private final AtomicLong referenceCount = new AtomicLong();
+
+  protected TreeLockNode(String name, LockManager manager) {
+    this.name = name;
```

Comment:
IF the class is thread safe, we don't need using thread safe object of ConcurrentHashMap and AtomicLong

Reply:
It's NOT thread-safe.